traits/observers/_has_traits_helpers.py

```diff
@@ -0,0 +1,114 @@
+"""
+
+from traits.ctraits import CHasTraits
+from traits.trait_base import Undefined, Uninitialized
+from traits.observers._exceptions import NotifierNotFound
+from traits.observers._observe import add_or_remove_notifiers
+
+
+#: List of values not to be passed onto the next observers because they are
+#: filled values for uninitialized/undefined traits, and no observables
+#: can be derived from them.
+UNOBSERVABLE_VALUES = [
+    Undefined,
+    Uninitialized,
+    None,
+]
+
+
+def object_has_named_trait(object, name):
+    """ Return true if a trait with the given name is defined on the object.
+
+    Parameters
+    ----------
+    object : any
+        Any object
+    name : str
+        Trait name to look for.
+
+    Returns
+    -------
+    boolean
+    """
+    return (
+        isinstance(object, CHasTraits) and object._trait(name, 0) is not None
+    )
+
+
+def iter_objects(object, name):
+    """ Yield the object referenced by the named trait, unless the value is not
+    defined or is one of the few unobservable filled values.
+
+    This function should not introduce side-effect (e.g. evaluating default)
+    on the given object.
+
+    Parameters
+    ----------
+    object : HasTraits
+        Object from which the trait value will be obtained.
+    name : str
+        Name of the trait.
+
+    Yields
+    ------
+    value : any
+        The value of the trait, if it is defined and is not one of those
+        skipped values.
+    """
+    value = object.__dict__.get(name, Undefined)
+    if value not in UNOBSERVABLE_VALUES:
+        yield value
+
+
+def observer_change_handler(event, graph, handler, target, dispatcher):
+    """ Maintain downstream notifiers when the trait changes.
+
+    Parameters
+    ----------
+    event : TraitObserverEvent
+        The event that triggers this function.
+    graph : ObserverGraph
+        Description for the *downstream* observers, i.e. excluding the observer
+        that contributed this maintainer function.
+    handler : callable
+        User handler.
+    target : object
+        Object seen by the user as the owner of the observer.
+    dispatcher : callable
+        Callable for dispatching the handler.
+    """
+    if event.old not in UNOBSERVABLE_VALUES:
+        try:
+            add_or_remove_notifiers(
+                object=event.old,
+                graph=graph,
+                handler=handler,
+                target=target,
+                dispatcher=dispatcher,
+                remove=True,
+            )
+        except NotifierNotFound:
+            # The old value could be a default value that does not have any
+            # notifier.
+            pass
+
+    if event.new not in UNOBSERVABLE_VALUES:
+        add_or_remove_notifiers(
+            object=event.new,
+            graph=graph,
+            handler=handler,
+            target=target,
+            dispatcher=dispatcher,
+            remove=False,
+        )
```

Comment:
Functions in these modules are refactored as they will be reused by `FilteredTraitObserver`.